contracts/contracts/strategies/ConvexEthMetaStrategy.sol

```diff
@@ -21,75 +21,141 @@ contract ConvexEthMetaStrategy is InitializableAbstractStrategy {
     using StableMath for uint256;
     using SafeERC20 for IERC20;
 
-    uint256 internal constant MAX_SLIPPAGE = 1e16; // 1%, same as the Curve UI
-    address internal constant ETH_ADDRESS =
+    uint256 public constant MAX_SLIPPAGE = 1e16; // 1%, same as the Curve UI
+    address public constant ETH_ADDRESS =
         0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE;
-    address internal cvxDepositorAddress;
-    IRewardStaking public cvxRewardStaker;
-    uint256 internal cvxDepositorPTokenId;
-    ICurveETHPoolV1 internal curvePool;
-    IERC20 internal lpToken;
-    IERC20 internal oeth;
-    IWETH9 internal weth;
+
+    // The following slots have been deprecated with immutable variables
+    // slither-disable-next-line constable-states
+    address private _deprecated_cvxDepositorAddress;
+    // slither-disable-next-line constable-states
+    address private _deprecated_cvxRewardStaker;
+    // slither-disable-next-line constable-states
+    uint256 private _deprecated_cvxDepositorPTokenId;
+    // slither-disable-next-line constable-states
+    address private _deprecated_curvePool;
+    // slither-disable-next-line constable-states
+    address private _deprecated_lpToken;
+    // slither-disable-next-line constable-states
+    address private _deprecated_oeth;
+    // slither-disable-next-line constable-states
+    address private _deprecated_weth;
+
+    // Ordered list of pool assets
+    // slither-disable-next-line constable-states
+    uint128 private _deprecated_oethCoinIndex = 1;
+    // slither-disable-next-line constable-states
+    uint128 private _deprecated_ethCoinIndex = 0;
+
+    // New immutable variables that must be set in the constructor
+    address public immutable cvxDepositorAddress;
+    IRewardStaking public immutable cvxRewardStaker;
+    uint256 public immutable cvxDepositorPTokenId;
+    ICurveETHPoolV1 public immutable curvePool;
+    IERC20 public immutable lpToken;
+    IERC20 public immutable oeth;
+    IWETH9 public immutable weth;
+
     // Ordered list of pool assets
-    uint128 internal oethCoinIndex;
-    uint128 internal ethCoinIndex;
+    uint128 public constant oethCoinIndex = 1;
+    uint128 public constant ethCoinIndex = 0;
 
-    // used to circumvent the stack too deep issue
-    struct InitializeConfig {
-        address curvePoolAddress; //Address of the Curve pool
+    /**
+     * @dev Verifies that the caller is the Strategist.
+     */
+    modifier onlyStrategist() {
+        require(
+            msg.sender == IVault(vaultAddress).strategistAddr(),
+            "Caller is not the Strategist"
+        );
+        _;
+    }
+
+    /**
+     * @dev Checks the Metapools balances have improved and the balances
+     * have not tipped to the other side.
+     */
+    modifier improveMetapoolBalance() {
```

Comment:
Can we do these checks on the normal deposit and withdraw methods?

Reply:
Yes, I think it's a good idea. Its another layer of protection if the vaultValueChecker wasn't used for some reason

The extra gas costs are
deposit 3,659
depositAll 3,662
withdraw  3,660
withdrawAll 3,663

```
|  Contract               ·  Method                              ·  Min        ·  Max        ·  Avg        ·  # calls      ·  eur (avg)  │
··························|······································|·············|·············|·············|···············|··············
|  ConvexEthMetaStrategy  ·  deposit                             ·          -  ·          -  ·    1571158  ·            4  ·      26.35  │
··························|······································|·············|·············|·············|···············|··············
|  ConvexEthMetaStrategy  ·  depositAll                          ·          -  ·          -  ·    1571433  ·            2  ·      26.35  │
··························|······································|·············|·············|·············|···············|··············
|  ConvexEthMetaStrategy  ·  withdraw                            ·          -  ·          -  ·    1570052  ·            4  ·      26.33  │
··························|······································|·············|·············|·············|···············|··············
|  ConvexEthMetaStrategy  ·  withdrawAll                         ·    1548363  ·    1553092  ·    1549939  ·            6  ·      25.99  │
```

```
|  Contract               ·  Method                              ·  Min        ·  Max        ·  Avg        ·  # calls      ·  eur (avg)  │
··························|······································|·············|·············|·············|···············|··············
|  ConvexEthMetaStrategy  ·  deposit                             ·          -  ·          -  ·    1574817  ·            4  ·      26.43  │
··························|······································|·············|·············|·············|···············|··············
|  ConvexEthMetaStrategy  ·  depositAll                          ·          -  ·          -  ·    1575095  ·            2  ·      26.43  │
··························|······································|·············|·············|·············|···············|··············
|  ConvexEthMetaStrategy  ·  withdraw                            ·          -  ·          -  ·    1573712  ·            4  ·      26.41  │
··························|······································|·············|·············|·············|···············|··············
|  ConvexEthMetaStrategy  ·  withdrawAll                         ·    1552026  ·    1556755  ·    1553602  ·            6  ·      26.07  │
```

Reply:
Thinking about is some more, I don't think we need them as we are adding liquidity to both sides. And for withdraws, we are removing proportionally so the balance does not change.

Reply:
Yes, in theory our math checks out, and we are safe, but I would love to have this guaranteed safeness wrapping this. It's beautiful when no matter how badly you bungle the code, nothing can break the invariant.

The extra gas isn't that bad for something that's used rarely like this.

Reply:
I'm wondering if this method actually works for the methods that add or remove from both sides at once. I think this only protects single adds or withdraws.

Reply:
I don't think it does. We can leave this off the existing methods, but let's make sure to comment that the check only works for single sided adds or removes.

Reply:
I went through the exact same thought process.

I've added a comment to the modifier.

BTW, I took Curve out of the modifier name so its now named `improvePoolBalance`